libsolidity/ast/Types.cpp

```diff
@@ -1828,12 +1807,35 @@ bool ArrayType::isDynamicallyEncoded() const
 	return isDynamicallySized() || baseType()->isDynamicallyEncoded();
 }
 
-bigint ArrayType::storageSizeUpperBound() const
+bigint ArrayType::sizeUpperBound(DataLocation _loc) const
 {
 	if (isDynamicallySized())
-		return 1;
-	else
+		return (_loc == DataLocation::Storage) ? 1 : memoryHeadSize();
+
+	if (_loc == DataLocation::Storage)
 		return length() * baseType()->storageSizeUpperBound();
```

Comment:
I think the code is easier to follow if you don't have to go look at `storageSizeUpperBound()` just to find out that it simply calls `sizeUpperBound(DataLocation::Storage)`:
```suggestion
	if (_loc == DataLocation::Storage)
		return length() * baseType()->sizeUpperBound(DataLocation::Storage);
```

Reply:
Done. If we decide for a universal `sizeUpperBound()` (despite mixed units), then `storageSizeUpperBound()` should be removed perhaps.